js_modules/dagit/packages/core/src/instance/InstanceConcurrency.tsx

```diff
@@ -0,0 +1,596 @@
+  };
+
+  const title = (
+    <>
+      Delete <Mono>{concurrencyKey}</Mono>
+    </>
+  );
+  return (
+    <Dialog isOpen={open} title={title} onClose={onClose}>
+      <DialogBody>
+        <Box flex={{justifyContent: 'flex-start'}}>
+          Delete concurrency limit&nbsp;<strong>{concurrencyKey}</strong>?
+        </Box>
+      </DialogBody>
+      <DialogFooter>
+        <Button intent="none" onClick={onClose}>
+          Close
+        </Button>
+        {isSubmitting ? (
+          <Button intent="primary" disabled>
+            Deleting...
+          </Button>
+        ) : (
+          <Button intent="danger" onClick={save}>
+            Delete limit
+          </Button>
+        )}
+      </DialogFooter>
+    </Dialog>
+  );
+};
+
+const ConcurrencyRunsDialog: React.FC<{
+  runIds?: string[];
+  title: string | React.ReactNode;
+  onClose: () => void;
+}> = ({runIds, onClose, title}) => {
+  const {data} = useQuery<RunsForConcurrencyKeyQuery, RunsForConcurrencyKeyQueryVariables>(
+    RUNS_FOR_CONCURRENCY_KEY_QUERY,
+    {
+      variables: {
+        limit: RUNS_LIMIT,
+        filter: {
+          runIds: runIds || [],
+        },
+      },
+      skip: !runIds || !runIds.length,
+    },
+  );
+
+  const [freeSlots] = useMutation<
+    FreeConcurrencySlotsForRunMutation,
+    FreeConcurrencySlotsForRunMutationVariables
+  >(FREE_CONCURRENCY_SLOTS_FOR_RUN_MUTATION);
+
+  const freeSlotsActionMenuItem = (run: RunTableRunFragment) => {
+    return doneStatuses.has(run.status)
+      ? [
+          <MenuItem
+            key="free-concurrency-slots"
+            icon="status"
+            text="Free concurrency slots for run"
+            onClick={async () => {
+              const resp = await freeSlots({variables: {runId: run.id}});
+              if (resp.data?.freeConcurrencySlotsForRun) {
+                await showSharedToaster({
+                  intent: 'success',
+                  icon: 'copy_to_clipboard_done',
+                  message: 'Freed concurrency slots',
+                });
+              }
+              onClose();
+            }}
+          />,
+        ]
+      : [];
+  };
+
+  return (
+    <Dialog
+      isOpen={!!runIds && runIds.length > 0}
+      title={title}
+      onClose={onClose}
+      style={{width: '60%', minWidth: '400px'}}
+    >
+      <Box padding={{vertical: 16}}>
+        {!data ? (
+          <Box padding={{vertical: 64}}>
+            <Spinner purpose="section" />
+          </Box>
+        ) : data.pipelineRunsOrError.__typename === 'Runs' ? (
+          <RunTable
+            runs={data.pipelineRunsOrError.results}
+            additionalActionsForRow={freeSlotsActionMenuItem}
+          />
+        ) : (
+          <Box padding={{vertical: 64}}>
+            <NonIdealState
+              icon="error"
+              title="Query Error"
+              description={
+                data.pipelineRunsOrError.__typename === 'PythonError'
+                  ? data.pipelineRunsOrError.message
```

Comment:
I'm not sure what this `message` value might look like, if it's going to be a whole stack trace or something -- in which case it might not fit nicely within a `NonIdealState`.